tests/test_helpers.py

```diff
@@ -515,6 +516,106 @@ def test_proxies_from_env_http_with_auth(mocker) -> None:
     assert proxy_auth.encoding == "latin1"
 
 
+# --------------------- get_env_proxy_for_url ------------------------------
+
+
+@pytest.mark.parametrize(
+    ("proxies", "url_input", "expected_err_msg"),
+    (
+        (
+            {"no_proxy": ["aiohttp.io"], "NO_PROXY": ["aiohttp.io"]},
+            "http://aiohttp.io/path",
+            r"Proxying is disallowed for `'aiohttp.io'`",
+        ),
+        (
+            {
+                "http_proxy": ["http://example.com"],
+                "HTTP_PROXY": ["http://example.com"],
+            },
+            "https://aiohttp.io/path",
+            r"No proxies found for `https://aiohttp.io/path` in the env",
+        ),
+        (
+            {
+                "https_proxy": ["https://example.com"],
+                "HTTPS_PROXY": ["https://example.com"],
+            },
+            "http://aiohttp.io/path",
+            r"No proxies found for `http://aiohttp.io/path` in the env",
+        ),
+        (
+            {},
+            "https://aiohttp.io/path",
+            r"No proxies found for `https://aiohttp.io/path` in the env",
+        ),
+        (
+            {
+                "https_proxy": ["https://example.com"],
+                "HTTPS_PROXY": ["https://example.com"],
+            },
+            "",
+            r"No proxies found for `` in the env",
+        ),
+    ),
+    ids=(
+        "url_matches_the_no_proxy_list",
+        "url_scheme_does_not_match_http_proxy_list",
+        "url_scheme_does_not_match_https_proxy_list",
+        "no_proxies_are_set",
+        "url_is_empty",
+    ),
+)
+def test_get_env_proxy_for_url_negative(
+    monkeypatch, proxies, url_input, expected_err_msg
+) -> None:
+    for schema in getproxies_environment().keys():
+        monkeypatch.delenv(schema, False)
+
+    for proxy_type, proxy_list in proxies.items():
+        monkeypatch.setenv(proxy_type, ",".join(proxy_list))
```

Comment:
`str.join()` never has any effect because all the params have exactly one item in the list. I'd like to make two points here:
1) If two proxies can be present, we should actually test this case: there's no point in implying that this is the case but not checking it.
2) In tests, we must do as few runtime-dependent changes / conditional coding as possible. The cases should be statically defined because adding clever code to tests opens up more possibilities to add bugs in those tests. They should be as simple as possible. So instead of having a list of one item that is concatenated into the same one item, it's better to (1) remove `str.join()`, (2) turn lists into strings and (3) just have a string with two proxies and a comma in between in the specific param.

Reply:
1. I used a list because the `xxx_proxy` can be a comma separated list. `no_proxy` is one that I see often with a list of domains/ips not to proxy. I'll add more test cases where there's multiple proxies in the list.
2. Okay, I'll make them a comma separated string.